Correctly compute baseline of wide-layout attachments

squelart opened the pull request:
#### ee150aed9da80fa92ec8a7c8aa24fedd46998392
<pre>
Correctly compute baseline of wide-layout attachments
<a href="https://bugs.webkit.org/show_bug.cgi?id=258868">https://bugs.webkit.org/show_bug.cgi?id=258868</a>
rdar://problem/111771870

Reviewed by Dean Jackson.

Previously it was falling back on the narrow/legacy renderer&apos;s baseline computation, which
happened to look okay.
This is fixed by doing a proper computation using a designated inner element, in this case
the image element because it has the same height&amp;baseline in all attachments (so that
attachments on the same line will be aligned regardless of their contents).

Also, because it was going through the legacy layout path, HTMLAttachmentElement::requestIconWithSize
triggered its debug-assertion that it should only be used with a narrow/legacy attachment; in release
it didn&apos;t crash but probably did an extra unnecessary icon request.

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::wideLayoutImageElement const):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/rendering/RenderAttachment.cpp:
(WebCore::RenderAttachment::baselinePosition const):

Canonical link: <a href="https://commits.webkit.org/265831@main">https://commits.webkit.org/265831@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2f22a78fcb6f4ec308ec7b0d9d07b60f57f03fa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12350 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12653 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13750 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11587 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12023 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14762 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12364 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14283 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12168 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12978 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10156 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14162 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10270 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10892 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18017 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11349 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11051 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14219 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11539 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9498 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10753 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2938 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15077 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11390 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/RenderAttachment.cpp:
```diff
@@ -97,8 +97,18 @@ void RenderAttachment::layout()
+        auto baselineElementTop = baselineElementRect.y() - attachmentRect.y();
+        if (auto* baselineElementRenderBox = baselineElement->renderBoxModelObject())
+            return LayoutUnit(baselineElementTop) + baselineElementRenderBox->baselinePosition(fontBaseline, firstLine, lineDirectionMode, linePositionMode);
+        return baselineElementTop + baselineElementRect.height();
```
One thing I'm wondering is if we should be using the margin box for the renderer here? I was looking at the other implementations of `baselinePosition` and they all see to return a value in terms of the margin box, but here we are using the border box.

sammygill replied:
(sorry for the noise I guess I didn't approve my review on GitHub when I first tried to add the comments)

squelart replied:
Thanks for noticing. I've filed follow-up https://bugs.webkit.org/show_bug.cgi?id=259172 to look into it.
It would only be a 1px difference, so I'm not too worried about it; but of course I'd prefer this to be correct eventually. Feel free to take it if you can't bear the thought of this egregious error surviving for long! 😄 (Or comment in the bug or radar if you feel it's more important than I realize.)

alanbaradlay commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -92,6 +92,7 @@ class HTMLAttachmentElement final : public HTMLElement {
 
     bool isWideLayout() const { return m_implementation == Implementation::WideLayout; }
     HTMLElement* wideLayoutShadowContainer() const { return m_containerElement.get(); }
+    HTMLElement* wideLayoutElementForBaseline() const;
```
please drop the "forBaseline" postfix. You are leaking layout concept to DOM for no obvious reason.

squelart replied:
But the whole point is that we need to pick a sub-element to be used as the baseline. Only the DOM knows what these sub-element mean and which one makes sense to use for the baseline. The renderer just sees a bag of elements and needs to be told which one should be used for the baseline.

I guess one solution would be to change this function to something like "wideLayoutImageElement", thereby exposing its meaning but not knowing it's for the baseline, and the renderer will pick this "image" thing for the baseline. 🤔

alanbaradlay replied:
how would you deal with the naming if this image element would be needed for some other task too?

squelart replied:
I'm not sure I understand the question.
- If you mean regarding the current "wideLayoutElementForBaseline" name, I could create a separate "wideElementFor_Task_" function that returns the same element for that other task.
- If you mean regarding the proposed name "wideLayoutImageElement", then the type of task doesn't matter anymore, another task could use that "image" for something else.
(Let me know if you meant something else.)

alanbaradlay commented on Source/WebCore/rendering/RenderAttachment.cpp:
```diff
@@ -97,8 +97,18 @@ void RenderAttachment::layout()
 {
+    if (auto* baselineElement = attachmentElement().wideLayoutElementForBaseline()) {
+        bool unusedIsReplaced;
+        auto attachmentRect = attachmentElement().pixelSnappedRenderRect(&unusedIsReplaced);
```
any particular reason why you consult the DOM tree for geometry (as opposed to the render tree)?

squelart replied:
The render code doesn't know what these sub-elements contain/mean.

alanbaradlay commented on Source/WebCore/rendering/RenderAttachment.cpp:
```diff
@@ -97,8 +97,18 @@ void RenderAttachment::layout()
+    if (auto* baselineElement = attachmentElement().wideLayoutElementForBaseline()) {
+        bool unusedIsReplaced;
+        auto attachmentRect = attachmentElement().pixelSnappedRenderRect(&unusedIsReplaced);
+        auto baselineElementRect = baselineElement->pixelSnappedRenderRect(&unusedIsReplaced);
```
any particular reason for using snapped geometry here?

squelart replied:
I copied other code blindly, I guess it's not needed here.

squelart committed:
Correctly compute baseline of wide-layout attachments
https://bugs.webkit.org/show_bug.cgi?id=258868
rdar://problem/111771870

Reviewed by Dean Jackson.

Previously it was falling back on the narrow/legacy renderer's baseline computation, which
happened to look okay.
This is fixed by doing a proper computation using a designated inner element, in this case
the image element because it has the same height&baseline in all attachments (so that
attachments on the same line will be aligned regardless of their contents).

Also, because it was going through the legacy layout path, HTMLAttachmentElement::requestIconWithSize
triggered its debug-assertion that it should only be used with a narrow/legacy attachment; in release
it didn't crash but probably did an extra unnecessary icon request.

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::wideLayoutImageElement const):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/rendering/RenderAttachment.cpp:
(WebCore::RenderAttachment::baselinePosition const):

Canonical link: https://commits.webkit.org/265831@main
